pyo3-derive-backend/src/pyenum.rs

```diff
@@ -0,0 +1,214 @@
+        }
+    }
+
+    impl_enum(&enum_.ident, variants)
+}
+
+fn impl_enum(
+    enum_: &syn::Ident,
+    variants: Vec<(syn::Ident, syn::ExprLit)>,
+) -> syn::Result<TokenStream> {
+    let enum_cls = impl_class(enum_, None)?;
+    let variant_names: Vec<syn::Ident> = variants
+        .iter()
+        .map(|(ident, _)| variant_enumname(enum_, ident))
+        .collect::<syn::Result<Vec<_>>>()?;
+
+    let variant_cls = variants
+        .iter()
+        .map(|(ident, _)| impl_class(ident, Some(enum_)))
+        .collect::<syn::Result<Vec<_>>>()?;
+    let variant_consts = variants
+        .iter()
+        .map(|(ident, _)| impl_const(enum_, ident))
+        .collect::<syn::Result<Vec<_>>>()?;
+
+    let to_py = impl_to_py(enum_, &variants)?;
+    let from_py = impl_from_py(enum_, &variants)?;
+
+    Ok(quote! {
+
+        #enum_cls
+
+        #(
+            struct #variant_names;
+        )*
+
+        #(
+            #variant_cls
+        )*
+
+        #to_py
+        #from_py
+
+        #[pymethods]
+        impl #enum_ {
+            #(
+                #[allow(non_upper_case_globals)]
+                #variant_consts
+            )*
+        }
+
+    })
+}
+
+fn impl_from_py(
+    enum_: &syn::Ident,
+    variants: &Vec<(syn::Ident, syn::ExprLit)>,
+) -> syn::Result<TokenStream> {
+    let matches = variants
+        .iter()
+        .map(|(ident, _)| {
+            variant_enumname(enum_, ident).map(|cls| {
+                let name = cls.to_string();
+                // TODO: this should be
+                // #cls as pyo3::type_object::PyTypeObject>::type_object(py).compare()
+                // but I can't figure out how to get py inside extract()
+                quote! {
+                    if typ_name == #name {
+                        return Ok(#enum_::#ident)
+                    }
+                }
+            })
+        })
+        .collect::<syn::Result<Vec<_>>>()?;
+
+    let errormsg = format!("Could not convert to {} enum", enum_);
+
+    Ok(quote! {
+        impl pyo3::conversion::FromPyObject<'_> for #enum_ {
+            fn extract(ob: &pyo3::PyAny) -> pyo3::PyResult<Self> {
+                let typ: &pyo3::types::PyType = ob.extract();
+                let typ_name = typ.name();
+                #(
+                    #matches
+                )*
+
+                Err(pyo3::exceptions::PyValueError::into(#errormsg))
+            }
+        }
+    })
+}
+
+fn impl_to_py(
+    enum_: &syn::Ident,
+    variants: &Vec<(syn::Ident, syn::ExprLit)>,
+) -> syn::Result<TokenStream> {
+    let matches = variants
+        .iter()
+        .map(|(ident, _)| {
+            variant_enumname(enum_, ident).map(|cls| {
+                quote! {
+                    #enum_::#ident => <#cls as pyo3::type_object::PyTypeObject>::type_object(py).to_object(py),
+                }
+            })
+        })
+        .collect::<syn::Result<Vec<_>>>()?;
+
+    Ok(quote! {
+        impl pyo3::FromPy<#enum_> for pyo3::PyObject {
```

Comment:
With the rebase on master, you'll need to change this to `impl IntoPy<PyObject> for #enum`